_posts/2021/2021-08-02-dont-call-replacement-classes-new.md

```diff
@@ -0,0 +1,95 @@
+---
+title: "Don't call a new version of something 'new' when refactoring"
+description: "A opinionated naming subtlety"
+layout: article
+category: ruby
+image:
+  base: '2021/dont-call-replacement-classes-new'
+  alt: "New"
+  credit: Nick Fewings
+  source: "https://unsplash.com/photos/1SsUquHPNT8"
+
+---
+
+In long-lived applications, we often want to rewrite or refactor an approach that we have previously taken.
+
+Here's a, non-obvious, naming technique I've used during refactoring to ensure that the improved code ends up in well named classes and any old implementation is clearly marked for deletion.
+
+You might find this helps organise your work or communicate the intent of future changes with your teammates.
+
+
+## Instead of...
+
+...littering your codebase with the word "new":
+
+```ruby
+class NewDoThingJob << ActiveJob
+  # new implementation
+end
+```
+
+or
+
+```ruby
+class DoThingJobVersionTwo << ActiveJob
+  # new implementation
+end
+```
+
+
+## Choose...
+
+to rename or, better, namespace the existing code to mark it as deprecated. Then create the replacement version with the correct future naming.
+
+```ruby
+module ToBeRemoved
+  class DoThingJob << ActiveJob
+    # existing implementation
+  end
+end
+```
+
+```ruby
+class DoThingJob < ToBeRemoved::DoThingJob
+```
+
+Then change all the places the non-namespaced version is called. Then deploy and ensure that no jobs are still enqueued with the non-namespaced version.
```

Comment:
This part still isn't so clear. Maybe be more explicit about what is being changed where and to what?

Reply:
```suggestion
Where the non-namespace implementation is called, change it to call the implementation in the deprecated namespace. Then deploy and ensure that the application is working as expected and that the non-namespaced version of the code is not called or enqueued anywhere.
```